### PR TITLE
Remove Appointment duplication

### PR DESCRIPTION
**Pull request by Gweaton**
## What does this pull request do?

Removes duplication in `Appointment` data structures, by removing the `AppointmentDetails` class that was introduced as a shared interface that could represent both original `Appointment` (used by for Initial Assessment)  and `ActionPlanAppointment` classes.

There's a lot of file changes required due to the number of classes that relied on these fields, so to break the important changes, we now have the following structure:

- `AppointmentSchedulingDetails` - scheduling details about an
appointment e.g. time, appointment type, duration. This also replaces
the `AppointmentUpdate` interface that we sent to the
Interventions Service as it contained the same fields.
- Base `Appointment` - extends `AppointmentSchedulingDetails` and adds
additional `SessionFeedback` which felt like it didn't belong on the
`SchedulingDetails` model.
- `InitialAssessmentAppointment` - extends `Appointment` and adds an
`id`.
- `ActionPlanAppointment` - extends `Appointment` and adds a
`SessionNumber`.

It's worth noting that `AppointmentSchedulingDetails` has nullable fields, where there weren't necessarily nullable fields on the duplicate interfaces that this new interface replaces. This shouldn't be an issue, as we seem to have null checks for this behaviour.

## What is the intent behind these changes?

This change was motivated by the need to add a `SessionType` e.g. "Group" or "1:1" field, as I didn't want to add this to all the duplicate classes, and now felt like a good time to re-use the shared interfaces. It's been a bit difficult adding behaviour to these classes when there are a lot of different types of appointment, so I'm hoping it should be easier going forward.

### CODE REVIEW

File: server/models/appointment.ts [hunk 1]
```diff
@@ -2,11 +2,21 @@ import Address from './address'
 import { AppointmentDeliveryType } from './appointmentDeliveryType'
 import SessionFeedback from './sessionFeedback'
 
-export default interface Appointment {
+export interface InitialAssessmentAppointment extends Appointment {
```
Comment (Gweaton) [inline server/models/appointment.ts:5]: Changes to this file are the bulk of this commit - the rest is really updating the rest of the codebase to use these interfaces.

**Review by lawrence-forooghian — COMMENTED** [review]
The approach looks sensible to me, good stuff. I think I'd like the @ministryofjustice/interventions-service folks to weigh in on this, though. The current interfaces are as they are because I anticipated that initial assessment appointments were our new "correct" way of doing an appoinment API, and action plan appointments were an initial attempt at an appointment API, that eventually we would bring in line with initial assessment appointments.

I'm using the word "appointment" here to mean "something that has been scheduled with a date / time / location".

Here's how I understand the current state of the APIs that work with appointments:

1. `GET /action-plan/:id/appointments` returns a list of objects that aren't actually appointments; they're a combination of a session and some information about that session's appointment (if one has been booked).
2. `GET /sent-referral/:id/supplier-assessment` returns a supplier assessment object, which then contains an array of appointments (which is empty if none have been booked).

And I _thought_ that we were going to change action plan appointments so that they also worked with appointments and a separate session object. In which case our app would only need one `Appointment` interface.

tl;dr: The changes in this PR bring the app more in line with how the API works _now_, and further away from how I thought we were going to change it to work. _Are_ we going to change the API?

File: server/decorators/appointmentDecorator.test.ts [hunk 2]
```diff
@@ -1,13 +1,13 @@
 import AppointmentDecorator from './appointmentDecorator'
-import appointmentFactory from '../../testutils/factories/initialAssessmentAppointment'
+import initialAssessmentAppointmentFactory from '../../testutils/factories/initialAssessmentAppointment'
```
[the remark this answers is not in the document]
Comment (Gweaton) [reply server/decorators/appointmentDecorator.test.ts:2]: Have fixed this up and replaced in a bunch of other places too, thanks for spotting those.

**Comment by Gweaton** [comment]
@lawrence-forooghian good points above, and you're right about the `session` vs `appointment` being a bit muddied at the moment.

I don't think there are any big plans to refactor `Appointment`s on the API at the moment. It would definitely make sense to do, but with a lot of people being away now / over the next few weeks it might not be considered a high priority. Either way, I think this feels a bit clearer for representing how things are now.

I'm gonna tag @andrewburrowsmoj and @tomsmyers as other people I "consulted" when making this change to see if they've got any thoughts on the above.

**Comment by lawrence-forooghian** [comment]
Yeah, I think my main worry was the introduction of nullability to the date / time etc fields. Because it means that if we add new appointment-related pages, they'll all have to have nullability checks in them even though the nullability is just a side-effect of one particular appointment type that isn't really an appointment.

**Comment by lawrence-forooghian** [comment]
To clarify on the above, I see the existing `Appointment` as being the correct way to do appointments for all future things that have appointments, and `ActionPlanAppointment` as being a special case that should have no influence on future appointment-related things (in particular, nullability).

**Review by andrewburrowsmoj — APPROVED** [review]
Happy with these changes. Thanks for doing it!
